zerver/actions/users.py

```diff
@@ -94,6 +96,93 @@ def do_delete_user(user_profile: UserProfile, *, acting_user: Optional[UserProfi
         )
 
 
+def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
+    """
+    This is a version of do_delete_user which does not delete message content.
+
+    The code is a bit tricky, because we want to, at some point, call
+    user_profile.delete() to trigger cascade deletes to clean up the variety
+    of related objects - but we need to avoid the cascades affecting
+    messages sent by the user.
+    """
+    if user_profile.realm.is_zephyr_mirror_realm:
+        raise AssertionError("Deleting zephyr mirror users is not supported")
+
+    do_deactivate_user(user_profile, acting_user=None)
+
+    subscribed_huddle_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.HUDDLE
+        ).values_list("recipient_id", flat=True)
+    )
+    user_id = user_profile.id
+    personal_recipient = user_profile.recipient
+    realm = user_profile.realm
+    date_joined = user_profile.date_joined
+
+    with transaction.atomic():
+        # The strategy is that before calling user_profile.delete(), we need to
+        # reassign Messages  sent by the user to a dummy user, so that they don't
+        # get affected by CASCADE. We cannot yet create a dummy user with .id
+        # matching that of the user_profile, so the general scheme is:
+        # 1. We create a *temporary* dummy for the initial re-assignment of messages.
+        # 2. We delete the UserProfile.
+        # 3. We create a replacement dummy user with its id matching what the UserProfile had.
+        # 4. This is the intended, final replacement UserProfile, so we re-assign
+        #    the messages from step (1) to it and delete the temporary dummy.
+        random_token = secrets.token_hex(16)
+        temp_replacement_user = create_user(
+            email=f"temp_deleteduser{random_token}@{get_fake_email_domain(realm)}",
+            password=None,
+            realm=realm,
+            full_name=f"Deleted User {random_token}",
```

Comment:
Why is this name different from the one used below?

Reply:
No particularly strong reason - the idea was just for this to be distinct in its format from the final dummy user. This one is supposed to be temporary, so it's different in nature - and actually shouldn't exist for more than a moment. A practical effect in favor of this is that if there was a bug causing this object to persist that'd be more noticeable, and less confusing, than if this looks like the legitimate, final dummy.

Could change this to be consistent and same as the final dummy user just as well though.

Reply:
OK went with something similar with a `(temp)` in its name, so it'll be easier to tie these to which original deletion if one gets leaked somehow.